Migrate in-app gallery to Jetpack Compose

MHShetty opened the pull request.

MHShetty committed:
Add verification metadata for Jetpack Compose

muhomorr commented on app/src/main/java/app/grapheneos/camera/ui/composable/screen/GalleryScreen.kt:
```diff
@@ -0,0 +1,349 @@
+                            val editIntent = Intent(Intent.ACTION_EDIT).apply {
+                                setDataAndType(currentItem.uri, currentItem.mimeType())
+                                putExtra(Intent.EXTRA_STREAM, currentItem.uri)
+                                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
```
Why is `Intent.FLAG_GRANT_WRITE_URI_PERMISSION` added here? It wasn't used by the previous implementation for this intent.

MHShetty replied:
It was initially added to test whether the image updates after editing it from another app (and fixes around it). This issue still exists due to the limited options present in the library (to re-compose the image) after returning back to the app. I have added a separate toggle for in-place editing in the local commits. Will update this thread as soon as the main issue gets fixed and the changes are pushed

MHShetty replied:
Have updated the code to ensure things work as expected. A separate option is now present to support editing the original image while the behavior of the original options has been set as before

muhomorr commented on app/src/main/java/app/grapheneos/camera/ui/composable/screen/GalleryScreen.kt:
```diff
@@ -0,0 +1,349 @@
+                onExit()
+            }
+
+            capturedItems.clear()
```
Missing synchronization, and ordering between these 2 coroutines doesn't look right. There's no guarantee about which will finish first.

MHShetty replied:
We have added checks to avoid unexpected behavior irrespective of which finishes first (`latestItem != null && capturedItems.isEmpty()`). We could merge it into a single coroutine callback to keep it simple and avoid the need for synchronization altogether

muhomorr commented on app/src/main/java/app/grapheneos/camera/ui/composable/components/TopBarActions.kt:
```diff
@@ -0,0 +1,172 @@
+                        onDismissRequest = {
+                            dropDownMenuExpanded = false
+                        },
+                        modifier = Modifier.width(200.dp)
```
It's not guaranteed that screen width is over 200 dp, it can be less than that on smaller devices when screen scaling is used.

MHShetty replied:
Ah that makes sense, will update the thread once I push the changes for the same

MHShetty replied:
We're filling up a certain ratio of the screen width based on the screen in which the top bar is situated

MHShetty committed:
Include compose with required dependencies

MHShetty committed:
Make CapturedItem serializable and parcelize able

MHShetty committed:
Add method to attempt to delete CapturedItem

MHShetty committed:
Add string resources

MHShetty committed:
Add dialog composables

MHShetty committed:
Add common utils and extension methods

MHShetty committed:
Add theme and typography for app

MHShetty committed:
Add media preview composables

MHShetty committed:
Add serializer and navtypes for routes

MHShetty committed:
Add QuickTooltip composable

MHShetty committed:
Add generic and custom topbar composables

MHShetty committed:
Add compose implementation for Video Player

MHShetty committed:
Add CapturedItemsRepository to access and observe captured items

MHShetty committed:
Add Gallery screen composable, viewmodel and route

MHShetty committed:
Add Extended Gallery screen composable, viewmodel and route

MHShetty committed:
Add CameraApp composable
(represents the entry point of the composable app)

MHShetty committed:
Use CameraApp composable
(and setup CapturedItemsRepository for use)

MHShetty committed:
Remove system actionbar from in-app gallery

MHShetty committed:
Remove old unused code
(related to views framework)

muhomorr commented:
What's the purpose of new (de)serializers for CapturedItem and Uri? They can be (de)serialized already since they implement the Parcelable interface.

MHShetty commented:
Both the Kotlin Serializable and the new Parcelable interfaces are needed for using type-safe navigation/custom objects in nav types. Initially, we were passing CapturedItem and List<CapturedItem> to specific routes/screen as parameters, but we later switched to having an activity-level repository to manage the captured items which is a much better and cleaner approach.

The only serializer and parcelable implementation we are currently using in the app is VideoUriSerializer class and the Parcelable implemented along with the VideoUri class itself. I have kept the other classes just in case if we need them later within the camera app or the gallery app. Type-safe navigation isn't very well documented unfortunately, a lot of the limitations it has isn't very clearly mentioned, so having the other serializable and nav type classes as reference would be helpful for future use

muhomorr commented:
What I'm saying is that Kotlin (de)serializers can reuse the implementation of Parcelable (de)serializers instead of duplicating them.

muhomorr commented:
A better solution in this case would probably be to not use route objects at all, since they don't work well with Parcelables. There's an option to pass route arguments as a Bundle instead. Bundle supports Parcelables directly.

MHShetty commented:
Route objects help us to do type-safe navigation by ensuring the right parameters are called (with respect to their type, order and in avoiding null check/assertions), and also improving the code readability while performing navigate calls apart from the compile-time checks it provides. Since we have already written the code for the same, I feel it's better to go ahead with the standard type-safe navigation approach

MHShetty committed:
Use no action bar theme
(for InAppGallery and VideoPlayer activities)

MHShetty committed:
Add immersive experience in focus mode of InAppGallery

MHShetty committed:
Separate UI-specific state from view model

MHShetty committed:
Add missing color
(the color was removed in a different PR but is continued to being used here)

MHShetty committed:
Disable predictive back gesture for MainActivity
(currently causes unexpected UI behavior; can be dealt with while migrating the rest of the camera app to Compose)

MHShetty committed:
Push new gallery screen
(when clicking on an image in extended gallery)

MHShetty committed:
Replace search icon with grid icon
(in gallery FAB)

MHShetty committed:
Clear snackbar message on dispose

MHShetty committed:
Reduce controllerShowTimeoutMs of VideoPlayer

MHShetty committed:
Add verification metadata

MHShetty committed:
Add verification metadata
